### PR TITLE
Add sync.sh for dev workflow

### PR DESCRIPTION
**Pull request by naamunds**
This script can be used to sync both source code and packages.

cc: @maririos @weshaggard @ellismg @dagood

### CODE REVIEW

**Comment by mellinoe** [comment]
I didn't read through the whole script, but I had a general question: When should I use this? Right now, packages are restored before a full build (build.proj) and they are also restored before each individual project (msbuild on .csproj), so what would be the value of restoring everything up-front as a separate step? Are we going to disable the above two options and this is a step in that direction?

File: sync.sh [hunk 1]
```diff
@@ -0,0 +1,254 @@
+#!/usr/bin/env bash
+
+usage()
+{
+    echo "Usage: sync [-p] [-s] [--branch <branch>] [--merge] [--rebase]"
+    echo "Repository syncing script."
+    echo
+    echo "  -p         Restore all NuGet packages for the repository"
+    echo "  -s         Update source by fast-forwarding upstream -> origin -> local"
```
Comment (weshaggard) [inline sync.sh]: I think this should just do a fetch like we do in sync.cmd the default actually mutating any of my branch state. It definitely makes too many assumptions about the names of remotes and structure of the branch in its current state.

**Comment by weshaggard** [comment]
@mellinoe ignoring the additional source control mutation options, which I think need further discussion, the general idea of when people will use this is when they want to be able to get all the assets from the from the networking, both source and packages, and then iterate with build or individual builds without needing to hit the network at all. Right now it is just a step in that direction but I think we will eventually stop restoring during build and require people to run sync before they can do a build. This is how our internal build works today.

**Comment by ellismg** [comment]
Just my two cents here and I understand that most of my feedback is colored by me being upset with the model rather than @naamunds's implementation of it.  The `sync.cmd` workflow was sort of bearable in TFS because I found that syncing to an arbitrary commit was something I did infrequently.  In git, I'm much more likely to to git checkout some random commit and do stuff. I also really dislike having to learn a set of scripts that augment but don't replace git.

I get the idea that we want to make it easy to say: "I'm about to get on an airplane, please pull down stuff I might need so I don't have to hit the network".  For packages, I get that having a script is helpful (I was happy with `msbuild /t:RestorePackages`, but I understand that in the limit we can't even assume msbuild is around).  It seems really weird to push this weird source-sync model onto git however.  I'm always in offline mode with git, so why can't I use my regular workflow?  This also feels like it breaks folks that want to work from VS all the time, since they are going to use VS's built in git integration instead of this cmd.

If all this script did was call `init-tools.sh` and then `msbuild /t:RestorePackages`, I think it would be reasonable. If you want to `git fetch --all` to help me out by getting all the latest stuff, I guess that's fine as well.

My general feedback is I really like the idea of `im-about-to-get-on-an-airplane.sh` but I don't like adding new/different stuff for the common case where I have an internet connection.  Please just let me use my regular tools in that case.

**Comment by stephentoub** [comment]
+$.01 to @ellismg's $.02.

**Comment by weshaggard** [comment]
@ellismg @stephentoub that is pretty much where I landed with sync.cmd (i.e. pull packages and fetch all) I kind of agree that we should probably use the git tool itself for the other source control work because I don't really want to force people into one workflow or another.

**Comment by naamunds** [comment]
I'm fine with changing it to just do a git fetch.  I originally preferred to do an implementation without any source syncing, but I got feedback that we should have a well-defined dev workflow that is consistent with the TFS one, so I implemented this one based on FxKit, which updates branches similar to how sync in TFS works.  It seems that the sentiment has changed since then, which is fine with me.

Do you guys think I should make the functionality for updating branches optional, or just remove it altogether? I would prefer to remove it, because I don't think it would add much value (for the reasons already stated here), and it may require maintaining. On Windows, FxKit provides this functionality, so people who want it could just use that, or could copy its code over to Linux pretty easily because it's implemented as a bash script.

**Comment by weshaggard** [comment]
I would remove the optional branch features and just make this match sync.cmd.

**Comment by naamunds** [comment]
I've updated the code to match sync.cmd, and made a few minor tweaks to sync.cmd.

**Comment by weshaggard** [comment]
LGTM

**Comment by eerhardt** [comment]
I had a discussion with @naamunds about what the plans are for the dev workflow, and I realized something I didn't before. Are we really planning on removing the "restore" step during the build? And forcing people to _have_ to call this sync.cmd/sh command in order for build to be successful?

If so, I think this is the wrong approach for two reasons:
1. It breaks the "learnablility" of how to work in our repo. New developers should just need to clone our repo and type "build" and the build just works. They shouldn't need to walk through a repo specific step-by-step instruction of how to build the code. This should be the same between all of our repos: coreclr, corert corefx, cli, and other Microsoft repos: aspnet, nuget, etc. And honestly any open source repo.
2. It breaks the workflow of adding a new dependency while I'm coding. If I have to go back and run "sync.sh" again because I added a new dependency in a project.json file, that is broken IMO.

**Comment by naamunds** [comment]
@eerhardt In response to your "learnability" criticism: We want to have restoring packages be an optional part of the build, and I'm not sure that learning the option to enable or disable package restore on build.sh or msbuild would be any easier than learning to run sync.sh. The only case where I think having package restore be an option on build would be preferable from a learnability perspective is if the default is to restore packages, because then devs could have a usable workflow without having to learn the skip-packages-restore option; they would just be inconvenienced by needing to wait for package restore when they don't want it. But even then, a lot of people would probably want to learn the option to skip restore, so I'm not sure that it's a big difference in learnability.

I agree that your second criticism (needing to run sync.sh again after adding a new dependency) is a significant issue that we will have to consider, because it might be easy for devs to forget to do that. I think that if we do use the approach of restoring packages only in sync.sh, we'd probably want to at least have build.sh give the user a message that they need to run sync.sh if some packages are missing. I'm not sure how we'd implement that, though.

**Comment by weshaggard** [comment]
For the second point currently we are still do a restore if you build an individual project configuration although I'm debating on disabling that as well as it just adds to the instability and performance issues.  We really need to eliminate all the network traffic for our basic workflows and only hit the network when we are intentional about doing so. 

As for the general workflow today with dotnet cli you must restore before you can compile/publish which is the model we are moving towards here as well.

File: sync.cmd [hunk 2]
```diff
@@ -50,8 +51,8 @@ if [%src%] == [true] (
 
 if [%packages%] == [true] (
   echo Restoring all packages ...
-  echo msbuild.exe %~dp0build.proj /t:BatchRestorePackages /nologo /v:minimal /p:RestoreDuringBuild=true /flp:v=diag;Append;LogFile=%synclog% >> %synclog%
-  call msbuild.exe %~dp0build.proj /t:BatchRestorePackages /nologo /v:minimal /p:RestoreDuringBuild=true /flp:v=diag;Append;LogFile=%synclog%
+  echo msbuild.exe %~dp0build.proj /t:BatchRestorePackages /nologo /v:minimal /p:RestoreDuringBuild=true /flp:v=detailed;Append;LogFile=%synclog% >> %synclog%
+  call msbuild.exe %~dp0build.proj /t:BatchRestorePackages /nologo /v:minimal /p:RestoreDuringBuild=true /flp:v=detailed;Append;LogFile=%synclog%
```
Comment (dagood) [inline sync.cmd]: Why detailed rather than diagnostic? I've found the extra info useful before.

**Comment by eerhardt** [comment]
> @naamunds  - The only case where I think having package restore be an option on build would be preferable from a learnability perspective is if the default is to restore packages, because then devs could have a usable workflow without having to learn the skip-packages-restore option

That seems like a fine option to me. It has been the default for the entire time I've been on this project, and it has been working.

> @weshaggard - As for the general workflow today with dotnet cli you must restore before you can compile/publish which is the model we are moving towards here as well.

We've had discussions on the team for this and I logged an issue for that as well: https://github.com/dotnet/cli/issues/1334

I want `build` to work ALL the time. I don't want to get 10 minutes into a build, and discover that I was missing some package, so now I need to go run a different command that the tooling should have just run for me. (This is a very easy situation to get in to when you are switching branches in git, since different commits can and will have different project.json files.)

Another point that came to me over the weekend is that the default behavior of sync.cmd/sh is broken. If I say `sync.cmd` with no arguments, it runs a `git fetch --all` and then runs `BatchRestorePackages`. The issue is that `git fetch` doesn't update the working directory (i.e. the files on disk). So when `BatchRestorePackages` is run, it runs against the OLD project.json files, which will have out-of-date versions in them.  Then when I move to the branch/commit that I actually want to work from, I get the updated project.json files and now the packages I previously restored are wrong. So then I need to run `sync.cmd` AGAIN. This workflow doesn't work for me. I don't see a single case where I would use this command for git operations.

Overall, my feedback is this:
1. Keep the default simple; make the advanced possible.
2. Don't over-engineer a simple process.

**Comment by markwilkie** [comment]
As @weshaggard  said, " We really need to eliminate all the network traffic for our basic workflows and only hit the network when we are intentional about doing so. " and I think this is a key point.  It would be a shame if every time I typed 'build', it went to the network.

Further, remember that anyone can cobble together the dev workflow building blocks to create/customize the wf as they see fit.  It would be great to stay in a place where the steps are discrete and clear.  We've got into trouble in the past trying to make the one magical script that rules them all.  Better to have available the encapsulated major steps and then let the dev be in control of their own destiny.  Today, the dev WF has only a few steps - they are: clean, sync, build. (and kinda test)  Soon, we'll have package, (a better) test, and probably publish.  These steps (BTW) are the same ones we'll use for the official build.  That way the dev is doing the same locally as we're doing "officially".

This approach also gives better flexibility.  For example, the forthcoming improved 'test' script will have an option to test the private bits across the entire matrix in Azure via Helix.  Further, the edit/build loop becomes very fast with very few dependencies.

**Comment by eerhardt** [comment]
As long as there is a command/script at the root of the enlistment that does what the CI runs do, I will be happy. I would prefer if that command/script was named "build" since that is consistent with the rest of our repos. But if you have a better name in mind that's fine.
I would not want to have to run a command, wait a few minutes, run another command, wait a few minutes, run another command, etc. I want to kick off 1 command, go away and come back and it is done.

**Comment by naamunds** [comment]
As @eerhardt points out, the dev workflow should be to update the local working tree before restoring packages. Because we don't want the Sync script to update the local working tree, this implies that that should be done before the running the Sync script to restore packages, meaning that the workflow is:
1. Clean
2. Update working tree
3. Restore packages
4. Build

In light of this, I question the value of having Sync run `git fetch`.  It doesn't make sense to run that after updating the working tree, so I think the only reason we might want it is if we want to make the "Sync" portion of the workflow be:
1. `sync.sh -s` (to start the update-working-tree step)
2. Finish updating your working tree as you see fit.
3. `sync.sh -p` (to restore packages)

But then not only is `sync.sh -s` not adding much value by running a command that devs may or may not use in their process of updating the working tree (for example, I usually just run `git pull`), it also seems to make the Sync step confusing, especially because it would be so different from what we have in the TFS sync.cmd, which does actually update the working tree.

So, my proposal is to remove the "-s" option from the Sync script altogether, and just have this script provide an easy way to restore packages. As such, I think it would be best to rename it "RestorePackages.sh".

I also agree with Eric that we should continue to have a single script that does everything necessary to build the repo. This could mean having build.sh continue to restore packages by default, or it could mean making a new script that simply calls two or more other scripts (e.g. RestorePackges.sh and build.sh, where build.sh does not restore packages).

**Comment by weshaggard** [comment]
Again the primary purpose Sync.sh/cmd is to gather stuff from the network both packages and source history. Which is why I still think fetch is useful as part of sync and I wouldn't rename this to RestorePackages because it logically fits into the Sync stage which means "pull all my remote dependencies". I realize it is slightly different in the semantics of updating the working directory between git and TFS but that mostly due to the nature of the differences in the source control system. I realize that people working in git branches may jump between them often but my expectation is that most git branches are short lived, which means they aren't much different between each other. That means you wouldn't need to necessarily run sync in all cases when you switch branches, but it is something you may need to consider doing it really depends on the changes you are making and jumping around to. 

As for the request to have one command that does everything I agree that is nice and we probably should have some sort of option for doing that but I still believe there is value in having these steps be distinct so that people can iterate on one of them without needing to run the full chain. Right now build.cmd does a number of things, clean (optional if you pass a parameter), sync (packages), build (product and tests), and test. Those are logically different steps and we should have those steps be unique blocks in our workflow. We then can have something that chains them together for folks that want a consistent set of steps but we should also allow people to just iterate on one of those steps, in particular build and/or test steps.

@maririos it would be good to publish your dev workflow diagram somewhere in our open documentation which outlines the flow we are expecting.

**Comment by markwilkie** [comment]
Wes and I are in super agreement.  (not exactly sure what that means…grin)  I 2nd the notion that we need to publish our dev workflow docs in the open.

**Comment by stephentoub** [comment]
I just read through this thread, and I admit that I'm a bit confused about where things are headed.

Here's what I want:

1. To be able to run build.cmd/sh for the entire repo and msbuild for an individual project and have all the right things happen, including all necessary packages getting restored as necessary.  This should be the default, as it's the "easy button", it just works. We've already lost something here with msbuild for an individual project failing after a git clean due to the core tools getting wiped, and I keep bumping up against that.  I don't want yet another set of "you're missing XYZ... fail" to get in the way.

2. To be able to suppress touching the network during builds, "airplane mode" as @ellismg refers to it.  I should be able to say "don't touch the network at all because either I'm expecting it to fail or I believe you should be able to succeed with what's on the machine and I don't want to pay the cost of the network access".  This should be opt-in, and those folks that care can learn how to opt-in.  Most of the time I'd just use (1), but if I'm in a coffee shop with a bad network connection, or on an airplane, or whatever, I can use this.  Optionally: make (2) an automatic fallback for (1) if a network operation fails, if the network is unreachable, etc.

3. Since (2) exists, to be able to say "download everything I could possibly need now, based on the current state of my source".  This should _not_ git fetch by default.  It should restore everything I need based on what I currently have in my repo, and if I want to pull in newer code from the repo, I can do so with the git commands I know and love.  Otherwise, the script is going to need to support additional options that I'm going to have to learn to override these operations, and I could have just used git to begin with.  I have no problem with the script allowing opt-in to doing additional git commands, but it should not be the default.

4. No unnecessary ceremony.  The scripts shouldn't do anything with git by default; I know how to interact with git and GitHub, I can do so.  Who's to say that there's even anything beyond my local repo... I could have a completely local repo with no remotes at all.

**Comment by markwilkie** [comment]
In reference to any confusion to where this is headed, here’s a bit of context:

This effort to improve dev workflow has been underway off and on since middle of last year when the runtime was on fire due to many interacting dependencies which made for extremely unreliable and slow builds.  Since that time, there have been many many discussions, emails with wide audiences, brownbags, and documentation published to ensure we’re solving problems that really exist in a way that makes sense moving forward.  And indeed, the noise from our developers in both CoreCLR and CoreFx has gone down, and the stability/speed of the builds have gone up.  Of course, it does not mean things are perfect….   

Now for looking forward:

To good news here is that we're doing our best to not force much on anyone.  In other words, one of the key principles is that a dev should be able to get their work done in a way that is most comfortable for them.  (within reason of course…) We also are working hard to make sure these workflow steps provide good value.  But perhaps for some, that’s less true.  The key again – make sure the workflow does not get in the way.  Building one monolithic build-thing does not allow for that flexibility of course.

As I read through the items that you need Steve, my impression is that we’re largely ok.  That is to say, it seems to me that the majority of any miss-alignment is either semantical or philosophical - with some items that really should be taken care of thrown in of course.   However, perhaps I’m missing something larger and if so, let’s get on the same page for sure.  At the very least, your comments will help improve the workflow and make it that much more usable for the devs across .NET.  For example, taking care of the core tools problem you mentioned.

And finally, here's a copy of the goals we have to better describe where we'd like to end up.  Also, please note that we're not done yet as not all of this is completed.

**Environment/Setup:**
- Needs to be fast and simple
- Should be obvious when it’s not right
- No magic: explicit setup over “I don’t know what it’s doing”

**Local Build/Validation:**
- Consistent workflow across platform
- Local validation should be a normal/easy part of developing code, as well as being a prerequisite to checking in.
- Failures should be obvious, including cause
- Clear and separate steps (e.g. “sync”, “incremental build”, “validate”, etc)
- Steps should be fast
- Option should exist to validate on another platform/environment as a service

**Culture:**
- Encourage the right cultural
- Ensure the engineering system helps drive desired cultural

**Comment by naamunds** [comment]
I think we are all in agreement except for two things:
1. In what form should we provide an "easy button" script that chains together the workflow steps like build.cmd/sh does today, including restoring packages?
2. Should sync.cmd/sh include an option to run `git fetch` (which wouldn't run if no arguments were passed), or should it be removed altogether?

I think 1) is an important discussion to have, but it's not directly related to this PR, because it doesn't affect the nature of the sync.sh script. How about we continue that discussion elsewhere, either through email, a Skype meeting, or on a new GitHub issue? What would be the best channel?

For 2), I would prefer that sync.cmd/sh not include an option to do `git fetch`, and that we accordingly rename it to RestorePackages.  As you pointed out, Wes, the Sync stage is "pull all my remote dependencies."  This includes two components: local working tree and packages.  In the initial design and implementation of the workflow in TFS, the Sync script encapsulates both of these components. In Git, though, we've decided that it's not feasible for it to fully encapsulate the working tree component.

As I mentioned earlier, if we do implement sync.sh to include `git fetch` (which may or may not be the first step of a dev's update-working-tree process), the Sync workflow would look like:
1.  `sync.sh -s`  (to start the update-working-tree step)
2. Finish updating your working tree as you see fit.
3.  `sync.sh`  (to restore packages)

In comparison, the Sync workflow in TFS is:
1. `sync.sh` (to update the working tree and restore packages)

Or, if the user wants to do the working tree and packages components steps separately in TFS:
1. `sync.sh -s` (to update the working tree)
2. `sync.sh -p` (to restore packages)

I think it would be confusing that the default behaviors (just running `sync.sh`) are not the same between TFS and Git, and even more confusing that `sync.sh -s` is similar but incomplete in Git. Devs might wonder, "Why did this not update my working tree like it does in TFS?", or worse, assume that it did, and waste time figuring out later that it didn't. Further, I think it's not very valuable anyway, because devs could easily run `git fetch --all -p` if that is indeed part of their update-working-tree process, and then it's clearer what is happening, rather than having to look at the usage for the script, and hoping that the script hasn't been changed to run some other Git commands.

In short, I feel like the confusion we would introduce by having `sync.sh -s` call `git fetch --all p` significantly outweighs the value it would add.

**Comment by markwilkie** [comment]
These are good points Nate and we should look at them more closely after @maririos gets back.

Given that this workflow has already been widely adopted across CoreFx and CoreCLR for the windows WF, I'd like to submit this PR with 'git fetch' and with the 'sync' name..  That said, I'm more than open to continuing the discussion.  This PR at least gets us one important step closer to a more converged WF across platforms which is important.

**Comment by weshaggard** [comment]
I'm personally not hung up on doing `git fetch --all` but I'm confused about why people think it is bad. It does nothing put get the source history from all your remotes (assuming you have any) and puts them into your local git database. It doesn't change the working directory at all. Sure in some cases if you are swapping branches and you do a git pull and you feel you need to sync then the fetch might be a no-op or only pull down history from other branches, but that shouldn't be harmful in anyway. 

As @markwilkie said lets move forward with this these options and make adjustments as needed.

**Comment by eerhardt** [comment]
> but I'm confused about why people think it is bad.

My thoughts are that it isn't "bad" in the sense of being detrimental to my environment.

Instead, my concern is that we are exerting effort to write scripts that I, as a corefx developer, am not going to get any value out of. As @stephentoub mentions above, "I know how to interact with git and GitHub, I can do so."

I feel our efforts would be better spent on activities that makes corefx developers more productive. Some examples off the top of my head:

- making the build faster
- ensuring the "dev workflow" build works on all of our supported platforms
  - note this is different than how Jenkins runs our build, which has allowed CI to pass, but it not to work on my machine
- making tests easier to run
  - currently if I want to run a single test on linux or OSX, I need to find some long 'msbuild' alias command that allows me to type "msbuild /t:BuildAndTest" on the directory I want to run. (And I'm not sure this process is even written down somewhere officially.)
  - making it easier to run corefx tests against a locally built coreclr (for when I make changes to mscorlib)

I'm sure other devs have things they feel would make them more productive - ex. @stephentoub's issue on not being able to build an individual project after a "clean".

File: sync.cmd (hunk 2 above)
Comment (naamunds) [inline sync.cmd]: Using detailed verbosity is consistent with clean.sh/cmd.  I think it's a better default because it keeps the log smaller, and if it's insufficient for debugging purposes, users can manually run it with diagnostic verbosity.

Comment (dagood) [inline sync.cmd]: Fair enough, the logs do get large. Diagnostic vs. detailed shouldn't matter for package restore anyway, on second thought.
